### ISSUE TITLE
Implement CLI arguments for generate key algorithm selection

### ISSUE THREAD

**Issue by ni4** [riboseinc/rnp#486]
## Description

Currently we have only default (RSA) key generation and expert key generation mode. However the last one requires stdin input and has some disadvantages - like, it doesn't allow to generate ECDSA/ECDH keypair.

So following parameters/modifications are proposed to be added to `rnpkeys`:
- `--generate rsa [bits]`
- `--generate dsa [bits]`
- `--generate ecdsa [curve]`
- `--generate eddsa [curve]`
- `--generate sm2`
- `--subkey [rsa|dsa|ecdsa|ecdh|eddsa|sm2] [bits|curve]`
- `--no-subkey`

Currently we use `--generate-key` but shortened `--generate` clearly describes the command's purpose.
In this generation mode encrypting subkey will be added by default, unless disabled by `--no-subkey`, and have the same passphrase as the main key.

Also expert generation mode should be modified to ask for subkey's presense and allow to select subkey's algorithm/curve.

**Comment by ni4** [comment]
@dewyatt @ronaldtse What do you think? I added it mostly for easier enhancement of CLI tests however it could be appreciated by users as well. Within current architecture it should not take a lot of time for implementation.

**Comment by ronaldtse** [comment]
@ni4 I agree we should have at least the key generation algorithm accessible by CLI. 

I'll leave it for @dewyatt to comment on the CLI syntax. For `sm2` the argument should be same as `ecdsa` which allows for a curve selection.

**Comment by ni4** [comment]
@ronaldtse Thanks, added curve to sm2 key algorithm. Waiting for @dewyatt's and others comments.

**Comment by flowher** [comment]
nitpicking: I think when generating key for SM2, it should not to be possible to select a curve. AFAIK SM2 specifies a curve to be used.
Similarly for eddsa - currently we only use ed25519 - however in this case, it may change in future

**Comment by dewyatt** [comment]
For SM2, isn't it completely OK to select alternative curves, but it's just not (OSCCA etc) compliant? I feel like we had a discussion before, but can't find it - @ronaldtse will know.

**Comment by ronaldtse** [comment]
@dewyatt yes we should allow SM2 to use any specified curve. 👍

**Comment by flowher** [comment]
How about hash? Should we allow only SM3 or any?

**Comment by dewyatt** [comment]
> How about hash? Should we allow only SM3 or any?

It seems like yes. (That's the discussion I was thinking of before: https://github.com/riboseinc/rnp/pull/422#issuecomment-324526447)

**Comment by flowher** [comment]
Thnx. Also #436. I think we first need to have possibility to use SM2 with selectable hash function before implementing SM2 with any curve. First is a work to be done in botan.

**Comment by ni4** [comment]
Regarding hash functions : RFC 4880 in 13.6 describes hash usage/truncation for the DSA. I think we should follow this rule - so, every hash may be used with every PK algorithm. However, probably we should check whether they corresponds in security bits (like there is no reason to use SHA-512 with RSA-1024, or MD5 with RSA-4096)

**Comment by ronaldtse** [comment]
Agree with @ni4 that we should allow every hash algorithm with every PK algorithm. It could be beneficial to check security bits correspondence but it probably should only deliver a "warning"?

**Comment by flowher** [comment]
Regarding security side - some combinations are not possible, for example you can't use P-521 with SHA-256 in ECDSA (it just doesn't work by design). Point "15.  {14} Security Considerations" has more details on that. 

Probably for SM2 there will be similar combination which won't work (code as is today has some checks already: https://github.com/randombit/botan/blob/master/src/lib/pubkey/sm2/sm2.cpp#L174)
